### PR TITLE
Statistics backend integration

### PR DESCRIPTION
**Pull request by Tom-The-Geek**
Initial work on integrating with [the persistence backend](https://github.com/Tom-The-Geek/nucleoid-persistence-backend) for statistic storage.
On game close, the `GameSpace`'s `GameStatisticBundle`s are sent using the backend integrations connection to the backend server if configured.

Notes:
- This shouldn't be merged until at least NucleoidMC/plasmid#154 and NucleoidMC/nucleoid-backend#11 are merged

Questions:
- Should statistics only be sent when a game closes with `GameCloseReason.FINISHED`, or for any reason (eg. `CANCELLED`)?

### CODE REVIEW

File: src/main/java/xyz/nucleoid/extras/statistics/NucleoidStatistics.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    private final StatisticsBackendConfig config;
+
+    private NucleoidStatistics(StatisticsBackendConfig config) {
+        this.config = config;
+        this.attachEventListener();
+    }
+
+    private void attachEventListener() {
+        GameEvents.CLOSING.register((space, reason) -> {
+            // TODO: Should we only submit statistics on reason == FINISHED?
```
Comment (Gegy) [inline src/main/java/xyz/nucleoid/extras/statistics/NucleoidStatistics.java]: Regarding this- yep! Definitely only on `FINISHED`. 👍

File: src/main/java/xyz/nucleoid/extras/statistics/NucleoidStatistics.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        Util.getIoWorkerExecutor().execute(() -> {
+            LOGGER.debug("Submitting statistic bundle for '{}'...", namespace);
+
+            JsonObject body = new JsonObject();
+            body.addProperty("server_name", this.config.getServerName());
+            body.addProperty("namespace", namespace);
+            body.add("stats", bundle.encodeBundle());
+
+            try {
+                URL url = new URL(this.config.getBackendUrl() + "/stats/upload");
```
Comment (Gegy) [inline src/main/java/xyz/nucleoid/extras/statistics/NucleoidStatistics.java]: Hmm, should this be a separate HTTP endpoint or does including it within the existing backend mechanism make the most sense?

Comment (Tom-The-Geek) [reply src/main/java/xyz/nucleoid/extras/statistics/NucleoidStatistics.java]: Currently its a [separate backend service](https://github.com/Tom-The-Geek/nucleoid-persistence-backend), that isn't part of the normal backend, and so it just uses a HTTP request to send the bundle. I can work on integrating it into the current backend at some point, but for now it just uses a simple HTTP API.

Comment (Tom-The-Geek) [reply src/main/java/xyz/nucleoid/extras/statistics/NucleoidStatistics.java]: Now uses the integrations connection to upload statistics

**Comment by Tom-The-Geek** [comment]
Closing this as I'm going to move it to the 1.17 branch

**Pull request closed without merging**